Add Host OS package environments to ISO installation

olavphibm opened the pull request.

Commit:
Add Host OS package environments to ISO installation

Those environments contain groups of packages and appear to the user
in the "Software selection" screen, along with the default CentOS ones,
which are merged from another comps.xml available remotely. They must
contain the base packages groups that are installed in a "Minimal
install", along with a group containing our built packages.

Commit:
Create yum groups for each open-power-host-os metapackage we build

We previously reused the CentOS group "virtualization-hypervisor" by
adding our packages. We now keep the CentOS groups as is and only add
new ones for Host OS.
Alter group XML creation to allow the group name to be more human
readable.

olavphibm commented:
build iso

olavphibm commented:
start iso

alanoe commented on config.yaml:
```diff
@@ -40,6 +40,8 @@ build_iso:
     mock_args:
```
Regarding commit msg, pls explain explicitly that a "package environment" is a group of package groups.

alanoe commented on config.yaml:
```diff
@@ -40,6 +40,8 @@ build_iso:
     mock_args:
     packages_dir: "result/packages/latest"
     automated_install_file: "host-os.ks"
+    base_packages_groups:
```
I suggest "base_distro_minimal_packages_groups" or  base_distro_base_packages_groups" instead of "base_packages_groups". Also, document here:
- what an environment is 
- that this represents the base distro's minimal list of package groups which must be installed so that the OS works
- that these packages groups which be available in all package environments

alanoe commented on lib/config.py:
```diff
@@ -103,6 +103,9 @@
              default='host-os.ks'),
+    ('--base-packages-groups',):
+        dict(help='Base packages groups that should always be installed in '
+             'the target system', nargs='*'),
```
Base here means both "from base distro" and "minimal". Could you make this more explicit ?

alanoe commented on lib/packages_groups_xml_creator.py:
```diff
@@ -18,6 +18,20 @@
 
 COMPS_DOCTYPE = "<!DOCTYPE comps PUBLIC \"-//HostOS//DTD Comps info//EN\" \"comps.dtd\">"
 
+def convert_name_to_id(element_name, suffix):
```
minor side comment: appending a suffix is out of the scope of "converting a name to ID", but as this feature is reused, OK

alanoe commented on lib/packages_groups_xml_creator.py:
```diff
@@ -67,7 +99,39 @@ def create_group_xml(group_name, group_pkgs):
-def create_comps_xml(packages_groups):
+def create_environment_xml(environment_name, environment_groups):
+    """
+    Construct an XML object representing an 'environment' element.
```
You could explain explicitly that an environment is simply a named list of packages groups somewhere.

alanoe commented on lib/packages_groups_xml_creator.py:
```diff
@@ -67,7 +99,39 @@ def create_group_xml(group_name, group_pkgs):
+    Those elements will appear as options during the installation.
+
+    Args:
+        environment_name (str): name of an environment of package groups
```
If you explain what an  environment is above, IMHO you could simply use "packages environment" instead of "environment of packages groups" here.

alanoe commented on config.yaml:
```diff
@@ -41,34 +41,18 @@ build_iso:
+        OpenPOWER Host OS virtualization management:
+            - "open-power-host-os-virt-management"
+        OpenPOWER Host OS RAS:
+            - "open-power-host-os-ras"
```
if there will be a 1:1 relationship between package group and metapackages, couldn't we automate the creation of this list with code? You could keep the list of metapackages somewhere (or even better, assume they are the ones which start with open-power-host-os in release rpm spec in versions) and create the groups names from it.

Commit:
Rename "automated_install_packages_groups" option

The new name "iso_root_fs_packages_groups" makes its use clearer.
Update the option's doc string.

Commit:
Separate occurrences of "hostos" with underscore

Commit:
Improve COMPS_DOCTYPE string readability

And separate "HostOS" with a space.

olavphibm commented:
start build

alanoe commented on lib/config.py:
```diff
@@ -117,8 +117,9 @@
-        dict(help='Packages and packages groups which are installed using automated installation', nargs='*'),
+    ('--iso-root-fs-packages-groups',):
+        dict(help='Packages and packages groups that will be installed in '
+             'the ISO root file system.', nargs='*'),
```
These packages and packages groups belong to the base distro (CentOS), it seems? It'd be good to mention it somewhere.

olavphibm replied:
Mentioned that the groups passed to `--host-os-packages-groups` option are automatically added.

alanoe commented on config.yaml:
```diff
@@ -42,34 +42,18 @@ build_iso:
     packages_dir: "result/packages/latest"
```
minor: It'd be good to mention somewhere in commit msg that you are not only *creating* new groups, but *replacing* the old packages groups.

alanoe replied:
You are deleting the previous OP Host OS virtualization-hypervisor package group...

olavphibm replied:
Actually, this is a CentOS group that was modified by us, it will still exist.
Clarifying this is commit message.

alanoe commented on config.yaml:
```diff
@@ -41,7 +41,7 @@ build_iso:
     packages_dir: "result/packages/latest"
     automated_install_file: "host-os.ks"
-    host_os_packages_groups:
+    installable_environments:
```
why you removed "host_os" prefix? I'd prefer to differ the things which belong to Open Power Host OS and the ones which belong to Cent OS (base distro) eg the groups below

olavphibm replied:
Those environments do not have to contain Host OS packages. They do have CentOS groups too, coming from the `minimal_install_groups` option.
If, for some reason, someone asks us to include an installation option to install a set of CentOS packages from extras, or whatever, this could be used.

alanoe replied:
OK

alanoe commented on config.yaml:
```diff
@@ -54,7 +54,9 @@ build_iso:
             - "open-power-host-os-ras"
-    iso_root_fs_packages_groups::
+    minimal_install_groups:
+        - "core"
```
- I'd prefer to make it clear in var name this belongs to base distro (CentOS), hard to know these groups belongs to the base distro (Cent OS) and not OP Host OS without reading help below. Maybe you consider base_distro_minimal_install_groups is too long? Up to you.
- I'd prefer to refer to packages groups consistently. Why here core does not have a @ and below it does?

olavphibm replied:
- OK
- Splitted lists for groups and single packages (`grub2`) below to remove the confusing `@`s.

alanoe commented on lib/packages_groups_xml_creator.py:
```diff
@@ -82,21 +101,64 @@ def create_group_xml(group_name, group_pkgs):
+            intermediary packages group with the same name will be
+            created to contain those packages.
+        minimal_install_groups ([str]): packages groups that must be
+            present in every CentOS installation and will be added to
```
we are hardcoding "CentOS" instead of using "base distro (eg CentOS)"?

olavphibm commented:
start iso

Commit:
Split "iso_root_fs_packages_groups" option

Both packages and packages groups were passed through this option, via
the kickstart syntax that was not clear to the user. By splitting the
lists for packages groups and single packages, we are able to clarify
what they mean.

olavphibm commented:
start build

olavphibm commented:
start build
